File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.java
/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

package com.amplifyframework.predictions.aws.service;

import androidx.annotation.NonNull;

import com.amplifyframework.core.Consumer;
import com.amplifyframework.predictions.PredictionsException;
import com.amplifyframework.predictions.aws.AWSPredictionsPluginConfiguration;
import com.amplifyframework.predictions.aws.adapter.EntityTypeAdapter;
import com.amplifyframework.predictions.aws.adapter.SentimentTypeAdapter;
import com.amplifyframework.predictions.aws.adapter.SpeechTypeAdapter;
import com.amplifyframework.predictions.aws.configuration.InterpretTextConfiguration;
import com.amplifyframework.predictions.models.Entity;
import com.amplifyframework.predictions.models.EntityType;
import com.amplifyframework.predictions.models.KeyPhrase;
import com.amplifyframework.predictions.models.Language;
import com.amplifyframework.predictions.models.LanguageType;
import com.amplifyframework.predictions.models.Sentiment;
import com.amplifyframework.predictions.models.SentimentType;
import com.amplifyframework.predictions.models.SpeechType;
import com.amplifyframework.predictions.models.Syntax;
import com.amplifyframework.predictions.result.InterpretResult;
import com.amplifyframework.util.UserAgent;

import com.amazonaws.AmazonClientException;
import com.amazonaws.ClientConfiguration;
import com.amazonaws.auth.AWSCredentialsProvider;
import com.amazonaws.services.comprehend.AmazonComprehendClient;
import com.amazonaws.services.comprehend.model.DetectDominantLanguageRequest;
import com.amazonaws.services.comprehend.model.DetectDominantLanguageResult;
import com.amazonaws.services.comprehend.model.DetectEntitiesRequest;
import com.amazonaws.services.comprehend.model.DetectEntitiesResult;
import com.amazonaws.services.comprehend.model.DetectKeyPhrasesRequest;
import com.amazonaws.services.comprehend.model.DetectKeyPhrasesResult;
import com.amazonaws.services.comprehend.model.DetectSentimentRequest;
import com.amazonaws.services.comprehend.model.DetectSentimentResult;
import com.amazonaws.services.comprehend.model.DetectSyntaxRequest;
import com.amazonaws.services.comprehend.model.DetectSyntaxResult;
import com.amazonaws.services.comprehend.model.DominantLanguage;
import com.amazonaws.services.comprehend.model.PartOfSpeechTag;
import com.amazonaws.services.comprehend.model.SentimentScore;

import java.util.ArrayList;
import java.util.List;

/**
 * Predictions service for performing text interpretation.
 */
final class AWSComprehendService {
    private static final int PERCENT = 100;

    private final AmazonComprehendClient comprehend;
    private final AWSPredictionsPluginConfiguration pluginConfiguration;

    AWSComprehendService(
            @NonNull AWSPredictionsPluginConfiguration pluginConfiguration,
            @NonNull AWSCredentialsProvider credentialsProvider) {
        this.comprehend = createComprehendClient(credentialsProvider);
        this.pluginConfiguration = pluginConfiguration;
    }

    private AmazonComprehendClient createComprehendClient(@NonNull AWSCredentialsProvider credentialsProvider) {
        ClientConfiguration configuration = new ClientConfiguration();
        configuration.setUserAgent(UserAgent.string());
        return new AmazonComprehendClient(credentialsProvider, configuration);
    }

    void comprehend(
            @NonNull String text,
            @NonNull Consumer<InterpretResult> onSuccess,
            @NonNull Consumer<PredictionsException> onError
    ) {
        try {
            // First obtain the dominant language to begin analysis
            final Language dominantLanguage = fetchPredominantLanguage(text);
            final LanguageType language = dominantLanguage.getValue();

            // Actually analyze text in the context of dominant language
            final Sentiment sentiment = fetchSentiment(text, language);
            final List<KeyPhrase> keyPhrases = fetchKeyPhrases(text, language);
            final List<Entity> entities = fetchEntities(text, language);
            final List<Syntax> syntax = fetchSyntax(text, language);

            onSuccess.accept(InterpretResult.builder()
                    .language(dominantLanguage)
                    .sentiment(sentiment)
                    .keyPhrases(keyPhrases)
                    .entities(entities)
                    .syntax(syntax)
                    .build());
        } catch (PredictionsException exception) {
            onError.accept(exception);
        }
    }

    private Language fetchPredominantLanguage(String text) throws PredictionsException {
        // Language is a required field for other detections.
        // Always fetch language regardless of what configuration says.
        isResourceConfigured(InterpretTextConfiguration.InterpretType.LANGUAGE);

        DetectDominantLanguageRequest request = new DetectDominantLanguageRequest()
                .withText(text);

        // Detect dominant language from given text via AWS Comprehend
        final DetectDominantLanguageResult result;
        try {
            result = comprehend.detectDominantLanguage(request);
        } catch (AmazonClientException serviceException) {
            throw new PredictionsException(
                    "AWS Comprehend encountered an error while detecting dominant language.",
                    serviceException,
                    "See attached service exception for more details."
            );
        }

        // Find the most dominant language from the list
        DominantLanguage dominantLanguage = null;
        for (DominantLanguage language : result.getLanguages()) {
            if (dominantLanguage == null
                    || language.getScore() > dominantLanguage.getScore()) {
                dominantLanguage = language;
            }
        }

        // Confirm that there was at least one detected language
        if (dominantLanguage == null) {
            throw new PredictionsException(
                    "AWS Comprehend did not detect any dominant language.",
                    "Please verify the integrity of text being analyzed."
            );
        }

        String languageCode = dominantLanguage.getLanguageCode();
        LanguageType language = LanguageType.from(languageCode);
        Float score = dominantLanguage.getScore();

        return Language.builder()
                .value(language)
                .confidence(score * PERCENT)
                .build();
    }

    private Sentiment fetchSentiment(String text, LanguageType language) throws PredictionsException {
        // Skip if configuration specifies NOT sentiment
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.SENTIMENT)) {
            return null;
        }

        DetectSentimentRequest request = new DetectSentimentRequest()
                .withText(text)
                .withLanguageCode(language.getLanguageCode());

        // Detect sentiment from given text via AWS Comprehend
        final DetectSentimentResult result;
        try {
            result = comprehend.detectSentiment(request);
        } catch (AmazonClientException serviceException) {
            throw new PredictionsException(
                    "AWS Comprehend encountered an error while detecting sentiment.",
                    serviceException,
                    "See attached service exception for more details."
            );
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        String comprehendSentiment = result.getSentiment();
        SentimentScore sentimentScore = result.getSentimentScore();
        SentimentType predominantSentiment = SentimentTypeAdapter.fromComprehend(comprehendSentiment);
        final float score;
        switch (predominantSentiment) {
            case POSITIVE:
                score = sentimentScore.getPositive();
                break;
            case NEGATIVE:
                score = sentimentScore.getNegative();
                break;
            case NEUTRAL:
                score = sentimentScore.getNeutral();
                break;
            case MIXED:
                score = sentimentScore.getMixed();
                break;
            default:
                score = 0f;
        }

        return Sentiment.builder()
                .value(predominantSentiment)
                .confidence(score * PERCENT)
                .build();
    }

    private List<KeyPhrase> fetchKeyPhrases(String text, LanguageType language) throws PredictionsException {
        // Skip if configuration specifies NOT key phrase
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.KEY_PHRASES)) {
            return null;
        }

        DetectKeyPhrasesRequest request = new DetectKeyPhrasesRequest()
                .withText(text)
                .withLanguageCode(language.getLanguageCode());

        // Detect key phrases from given text via AWS Comprehend
        final DetectKeyPhrasesResult result;
        try {
            result = comprehend.detectKeyPhrases(request);
        } catch (AmazonClientException serviceException) {
            throw new PredictionsException(
                    "AWS Comprehend encountered an error while detecting key phrases.",
                    serviceException,
                    "See attached service exception for more details."
            );
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        List<KeyPhrase> keyPhrases = new ArrayList<>();
        for (com.amazonaws.services.comprehend.model.KeyPhrase comprehendKeyPhrase : result.getKeyPhrases()) {
            KeyPhrase amplifyKeyPhrase = KeyPhrase.builder()
                    .value(comprehendKeyPhrase.getText())
                    .confidence(comprehendKeyPhrase.getScore() * PERCENT)
                    .targetText(comprehendKeyPhrase.getText())
                    .startIndex(comprehendKeyPhrase.getBeginOffset())
                    .build();
            keyPhrases.add(amplifyKeyPhrase);
        }

        return keyPhrases;
    }

    private List<Entity> fetchEntities(String text, LanguageType language) throws PredictionsException {
        // Skip if configuration specifies NOT entities
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.ENTITIES)) {
            return null;
        }

        DetectEntitiesRequest request = new DetectEntitiesRequest()
                .withText(text)
                .withLanguageCode(language.getLanguageCode());

        // Detect entities from given text via AWS Comprehend
        final DetectEntitiesResult result;
        try {
            result = comprehend.detectEntities(request);
        } catch (AmazonClientException serviceException) {
            throw new PredictionsException(
                    "AWS Comprehend encountered an error while detecting entities.",
                    serviceException,
                    "See attached service exception for more details."
            );
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        List<Entity> entities = new ArrayList<>();
        for (com.amazonaws.services.comprehend.model.Entity comprehendEntity : result.getEntities()) {
            EntityType entityType = EntityTypeAdapter.fromComprehend(comprehendEntity.getType());
            Entity amplifyEntity = Entity.builder()
                    .value(entityType)
                    .confidence(comprehendEntity.getScore() * PERCENT)
                    .targetText(comprehendEntity.getText())
                    .startIndex(comprehendEntity.getBeginOffset())
                    .build();
            entities.add(amplifyEntity);
        }

        return entities;
    }

    private List<Syntax> fetchSyntax(String text, LanguageType language) throws PredictionsException {
        // Skip if configuration specifies NOT syntax
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.SYNTAX)) {
            return null;
        }

        DetectSyntaxRequest request = new DetectSyntaxRequest()
                .withText(text)
                .withLanguageCode(language.getLanguageCode());

        // Detect syntax from given text via AWS Comprehend
        final DetectSyntaxResult result;
        try {
            result = comprehend.detectSyntax(request);
        } catch (AmazonClientException serviceException) {
            throw new PredictionsException(
                    "AWS Comprehend encountered an error while detecting syntax.",
                    serviceException,
                    "See attached service exception for more details."
            );
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        List<Syntax> syntaxTokens = new ArrayList<>();
        for (com.amazonaws.services.comprehend.model.SyntaxToken comprehendSyntax : result.getSyntaxTokens()) {
            PartOfSpeechTag comprehendPartOfSpeech = comprehendSyntax.getPartOfSpeech();
            SpeechType partOfSpeech = SpeechTypeAdapter.fromComprehend(comprehendPartOfSpeech.getTag());
            Syntax amplifySyntax = Syntax.builder()
                    .id(comprehendSyntax.getTokenId().toString())
                    .value(partOfSpeech)
                    .confidence(comprehendPartOfSpeech.getScore() * PERCENT)
                    .targetText(comprehendSyntax.getText())
                    .startIndex(comprehendSyntax.getBeginOffset())
                    .build();
            syntaxTokens.add(amplifySyntax);
        }

        return syntaxTokens;
    }

    private boolean isResourceConfigured(InterpretTextConfiguration.InterpretType type) throws PredictionsException {
        // Check if text interpretation is configured
        InterpretTextConfiguration.InterpretType configuredType =
                pluginConfiguration.getInterpretTextConfiguration().getType();

        if (InterpretTextConfiguration.InterpretType.ALL.equals(configuredType)) {
            // ALL catches every type
            return true;
        } else {
            // Otherwise check to see if they are equal
            return configuredType.equals(type);
        }
    }

    @NonNull
    AmazonComprehendClient getClient() {
        return comprehend;
    }
}

File: aws-predictions/src/main/java/com/amplifyframework/predictions/aws/service/AWSComprehendService.kt
/*
 * Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 *  http://aws.amazon.com/apache2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
package com.amplifyframework.predictions.aws.service

import aws.sdk.kotlin.services.comprehend.ComprehendClient
import aws.sdk.kotlin.services.comprehend.model.DominantLanguage
import aws.sdk.kotlin.services.comprehend.model.LanguageCode
import aws.sdk.kotlin.services.comprehend.model.SyntaxLanguageCode
import aws.smithy.kotlin.runtime.auth.awscredentials.CredentialsProvider
import com.amplifyframework.core.Consumer
import com.amplifyframework.predictions.PredictionsException
import com.amplifyframework.predictions.aws.AWSPredictionsPluginConfiguration
import com.amplifyframework.predictions.aws.adapter.EntityTypeAdapter
import com.amplifyframework.predictions.aws.adapter.SentimentTypeAdapter
import com.amplifyframework.predictions.aws.adapter.SpeechTypeAdapter
import com.amplifyframework.predictions.aws.configuration.InterpretTextConfiguration
import com.amplifyframework.predictions.models.Entity
import com.amplifyframework.predictions.models.EntityType
import com.amplifyframework.predictions.models.KeyPhrase
import com.amplifyframework.predictions.models.Language
import com.amplifyframework.predictions.models.LanguageType
import com.amplifyframework.predictions.models.Sentiment
import com.amplifyframework.predictions.models.SentimentType
import com.amplifyframework.predictions.models.Syntax
import com.amplifyframework.predictions.result.InterpretResult
import java.util.ArrayList
import java.util.concurrent.Executors
import kotlinx.coroutines.runBlocking

/**
 * Predictions service for performing text interpretation.
 */
internal class AWSComprehendService(
    private val pluginConfiguration: AWSPredictionsPluginConfiguration,
    private val authCredentialsProvider: CredentialsProvider
) {
    val client: ComprehendClient = ComprehendClient {
        this.region = pluginConfiguration.defaultRegion
        this.credentialsProvider = authCredentialsProvider
    }

    private val executor = Executors.newCachedThreadPool()

    companion object {
        private const val PERCENT = 100
    }

    fun comprehend(
        text: String,
        onSuccess: Consumer<InterpretResult>,
        onError: Consumer<PredictionsException>
    ) {
        execute(
            {
                // First obtain the dominant language to begin analysis
                val dominantLanguage = fetchPredominantLanguage(text)
                val language = dominantLanguage.value

                // Actually analyze text in the context of dominant language
                val sentiment = fetchSentiment(text, language)
                val keyPhrases = fetchKeyPhrases(text, language)
                val entities = fetchEntities(text, language)
                val syntax = fetchSyntax(text, language)
                InterpretResult.builder()
                    .language(dominantLanguage)
                    .sentiment(sentiment)
                    .keyPhrases(keyPhrases)
                    .entities(entities)
                    .syntax(syntax)
                    .build()
            },
            { throwable ->
                PredictionsException(
                    "AWS Comprehend encountered an error while interpreting text.",
                    throwable,
                    "See attached exception for more details."
                )
            },
            onSuccess,
            onError
        )
    }

    @Throws(PredictionsException::class)
    private suspend fun fetchPredominantLanguage(text: String): Language {
        // Language is a required field for other detections.
        // Always fetch language regardless of what configuration says.
        isResourceConfigured(InterpretTextConfiguration.InterpretType.LANGUAGE)

        // Detect dominant language from given text via AWS Comprehend
        val result = try {
            client.detectDominantLanguage {
                this.text = text
            }
        } catch (exception: Exception) {
            throw PredictionsException(
                "AWS Comprehend encountered an error while detecting dominant language.",
                exception,
                "See attached exception for more details."
            )
        }

        // Find the most dominant language from the list
        var dominantLanguage: DominantLanguage? = null
        result.languages?.forEach { language ->
            val dominantLanguageScore = dominantLanguage?.score
            val currentLanguageScore = language.score
            if (dominantLanguage == null) {
                dominantLanguage = language
            } else if (dominantLanguageScore != null && currentLanguageScore != null) {
                dominantLanguage = if (currentLanguageScore > dominantLanguageScore) {
                    language
                } else {
                    dominantLanguage
                }
            }
        }

        // Confirm that there was at least one detected language
        if (dominantLanguage == null) {
            throw PredictionsException(
                "AWS Comprehend did not detect any dominant language.",
                "Please verify the integrity of text being analyzed."
            )
        }
        val languageCode = dominantLanguage!!.languageCode
        val language = LanguageType.from(languageCode)
        val score = dominantLanguage!!.score
        val languageBuilder = Language.builder()
            .value(language)
        if (score != null) {
            languageBuilder.confidence(score * PERCENT)
        }
        return languageBuilder.build()
    }

    @Throws(PredictionsException::class)
    private suspend fun fetchSentiment(text: String, language: LanguageType): Sentiment? {
        // Skip if configuration specifies NOT sentiment
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.SENTIMENT)) {
            return null
        }
        // Detect sentiment from given text via AWS Comprehend
        val result = try {
            client.detectSentiment {
                this.text = text
                this.languageCode = LanguageCode.fromValue(language.languageCode)
            }
        } catch (exception: Exception) {
            throw PredictionsException(
                "AWS Comprehend encountered an error while detecting sentiment.",
                exception,
                "See attached exception for more details."
            )
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        val comprehendSentiment = result.sentiment
        val sentimentScore = result.sentimentScore
        val predominantSentiment = SentimentTypeAdapter.fromComprehend(comprehendSentiment.toString())
        val score = when (predominantSentiment) {
            SentimentType.POSITIVE -> sentimentScore?.positive
            SentimentType.NEGATIVE -> sentimentScore?.negative
            SentimentType.NEUTRAL -> sentimentScore?.neutral
            SentimentType.MIXED -> sentimentScore?.mixed
            else -> 0f
        }
        if (score != null) {
            return Sentiment.builder()
                .value(predominantSentiment)
                .confidence(score * PERCENT)
                .build()
        }
        return null
    }

    @Throws(PredictionsException::class)
    private suspend fun fetchKeyPhrases(text: String, language: LanguageType): List<KeyPhrase>? {
        // Skip if configuration specifies NOT key phrase
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.KEY_PHRASES)) {
            return null
        }
        // Detect key phrases from given text via AWS Comprehend
        val result = try {
            client.detectKeyPhrases {
                this.text = text
                this.languageCode = LanguageCode.fromValue(language.languageCode)
            }
        } catch (exception: Exception) {
            throw PredictionsException(
                "AWS Comprehend encountered an error while detecting key phrases.",
                exception,
                "See attached exception for more details."
            )
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        val keyPhrases: MutableList<KeyPhrase> = ArrayList()
        result.keyPhrases?.forEach { comprehendKeyPhrase ->
            val keyPhraseText = comprehendKeyPhrase.text
            val keyPhraseScore = comprehendKeyPhrase.score
            val keyPhraseOffset = comprehendKeyPhrase.beginOffset
            if (keyPhraseText != null && keyPhraseScore != null && keyPhraseOffset != null) {
                val amplifyKeyPhrase = KeyPhrase.builder()
                    .value(keyPhraseText)
                    .confidence(keyPhraseScore * PERCENT)
                    .targetText(keyPhraseText)
                    .startIndex(keyPhraseOffset)
                    .build()
                keyPhrases.add(amplifyKeyPhrase)
            }
        }
        return keyPhrases
    }

    @Throws(PredictionsException::class)
    private suspend fun fetchEntities(text: String, language: LanguageType): List<Entity>? {
        // Skip if configuration specifies NOT entities
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.ENTITIES)) {
            return null
        }
        // Detect entities from given text via AWS Comprehend
        val result = try {
            client.detectEntities {
                this.text = text
                this.languageCode = LanguageCode.fromValue(language.languageCode)
            }
        } catch (exception: Exception) {
            throw PredictionsException(
                "AWS Comprehend encountered an error while detecting entities.",
                exception,
                "See attached exception for more details."
            )
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        val entities: MutableList<Entity> = ArrayList()
        result.entities?.forEach { comprehendEntity ->
            val entityType: EntityType =
                EntityTypeAdapter.fromComprehend(comprehendEntity.type.toString())
            val entityScore = comprehendEntity.score
            val entityText = comprehendEntity.text
            val entityOffset = comprehendEntity.beginOffset
            if (entityScore != null && entityText != null && entityOffset != null) {
                val amplifyEntity = Entity.builder()
                    .value(entityType)
                    .confidence(entityScore * PERCENT)
                    .targetText(entityText)
                    .startIndex(entityOffset)
                    .build()
                entities.add(amplifyEntity)
            }
        }
        return entities
    }

    @Throws(PredictionsException::class)
    private suspend fun fetchSyntax(text: String, language: LanguageType): List<Syntax>? {
        // Skip if configuration specifies NOT syntax
        if (!isResourceConfigured(InterpretTextConfiguration.InterpretType.SYNTAX)) {
            return null
        }
        // Detect syntax from given text via AWS Comprehend
        val result = try {
            client.detectSyntax {
                this.text = text
                this.languageCode = SyntaxLanguageCode.fromValue(language.languageCode)
            }
        } catch (exception: Exception) {
            throw PredictionsException(
                "AWS Comprehend encountered an error while detecting syntax.",
                exception,
                "See attached exception for more details."
            )
        }

        // Convert AWS Comprehend's detection result to Amplify-compatible format
        val syntaxTokens: MutableList<Syntax> = ArrayList()
        result.syntaxTokens?.forEach { comprehendSyntax ->
            val partOfSpeech = SpeechTypeAdapter.fromComprehend(comprehendSyntax.partOfSpeech?.tag.toString())
            val partOfSpeechScore = comprehendSyntax.partOfSpeech?.score
            val syntaxText = comprehendSyntax.text
            val syntaxOffset = comprehendSyntax.beginOffset
            if (partOfSpeechScore != null && syntaxText != null && syntaxOffset != null) {
                val amplifySyntax = Syntax.builder()
                    .id(comprehendSyntax.tokenId.toString())
                    .value(partOfSpeech)
                    .confidence(partOfSpeechScore * PERCENT)
                    .targetText(syntaxText)
                    .startIndex(syntaxOffset)
                    .build()
                syntaxTokens.add(amplifySyntax)
            }
        }
        return syntaxTokens
    }

    private fun isResourceConfigured(type: InterpretTextConfiguration.InterpretType): Boolean {
        // Check if text interpretation is configured
        val configuredType: InterpretTextConfiguration.InterpretType =
            pluginConfiguration.interpretTextConfiguration.type
        return if (InterpretTextConfiguration.InterpretType.ALL == configuredType) {
            // ALL catches every type
            true
        } else {
            // Otherwise check to see if they are equal
            configuredType == type
        }
    }

    private fun <T : Any> execute(
        runnableTask: suspend () -> T,
        errorTransformer: (Throwable) -> PredictionsException,
        onResult: Consumer<T>,
        onError: Consumer<PredictionsException>
    ) {
        executor.execute {
            try {
                runBlocking {
                    val result = runnableTask()
                    onResult.accept(result)
                }
            } catch (error: Throwable) {
                val predictionsException = if (error is PredictionsException) {
                    error
                } else {
                    errorTransformer.invoke(error)
                }
                onError.accept(predictionsException)
            }
        }
    }
}
